nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/CopyAzureBlobStorage_v12.java

```diff
@@ -0,0 +1,388 @@
+        BlobContainerSasPermission permissions = new BlobContainerSasPermission().setCreatePermission(true).setWritePermission(true).setAddPermission(true).setReadPermission(true);
+        OffsetDateTime utc = OffsetDateTime.now(ZoneOffset.UTC);
+        OffsetDateTime expiryTime = utc.plus(CopyAzureBlobStorage_v12.GENERATE_SAS_EXPIRY_HOURS, ChronoUnit.HOURS);
+        BlobServiceSasSignatureValues signatureValues = new BlobServiceSasSignatureValues(expiryTime, permissions);
+        return fromContainerClient.generateSas(signatureValues);
+    }
+
+    private void copy(final BlobClient blobClient,
+                      final HttpAuthorization httpAuthorization,
+                      final String sourceUrl,
+                      long blobSize,
+                      final BlobRequestConditions sourceRequestConditions,
+                      final BlobRequestConditions destinationRequestConditions) {
+        BlockBlobClient blockBlobClient = blobClient.getBlockBlobClient();
+
+        // If the blob size is below the limit, we use the one-shot upload endpoint.
+        if (blobSize < MAX_UPLOAD_BLOB_BYTES_LONG) {
+            final BlobUploadFromUrlOptions options = new BlobUploadFromUrlOptions(sourceUrl);
+            if (httpAuthorization != null) {
+                options.setSourceAuthorization(httpAuthorization);
+            }
+            options.setSourceRequestConditions(sourceRequestConditions);
+            options.setDestinationRequestConditions(destinationRequestConditions);
+            blockBlobClient.uploadFromUrlWithResponse(options, null, Context.NONE);
+            return;
+        }
+
+        final DecimalFormat df = new DecimalFormat("0000000");
+        long offset = 0;
+        int blockId = 1;
+        final List<String> blockIds = new ArrayList<>();
+
+        // Upload each block sequentially until we've uploaded the entire blob.
+        while (true) {
+            long count = Math.min(blobSize - offset, MAX_STAGE_BLOCK_BYTES_LONG);
+            if (count == 0) break;
+
+            final String base64BlockId = Base64.getEncoder().encodeToString(df.format(blockId).getBytes());
+            BlockBlobStageBlockFromUrlOptions blockBlobStageBlockFromUrlOptions = new BlockBlobStageBlockFromUrlOptions(base64BlockId, sourceUrl);
+            blockBlobStageBlockFromUrlOptions.setSourceRange(new BlobRange(offset, count));
+
+            if (httpAuthorization != null) {
+                blockBlobStageBlockFromUrlOptions.setSourceAuthorization(httpAuthorization);
+            }
+            blockBlobStageBlockFromUrlOptions.setSourceRequestConditions(sourceRequestConditions);
+            final int statusCode = blockBlobClient.stageBlockFromUrlWithResponse(blockBlobStageBlockFromUrlOptions, null, Context.NONE).getStatusCode();
+            if (statusCode != 201) {
+                throw new ProcessException(String.format("Failed staging one or more blocks (status: %d)", statusCode));
```

Comment:
Recommend adjusting the wording to clarify the error code source.
```suggestion
                throw new ProcessException(String.format("Failed staging one or more blocks: HTTP %d", statusCode));
```